File: Week_02/PPMH_BT7/BT66/PPMH_BT66.py
def implies(a, b):
    return (not a) or b

def biconditional(a, b):
    return a == b;

def main():
    # a
    print("a")
    sat = False
    for i in range(16):
        p = (i // 8) % 2 == 1
        q = (i // 4) % 2 == 1
        r = (i // 2) % 2 == 1
        s = i % 2 == 1

        value = ((p or q or not r)
            and (p or not q or not s)
            and (p or not r or not s)
            and (not p or not q or s)
            and (p or q or not s))
        
        if value:
            print(f"Satisfiable for p={int(p)}, q={int(q)}, r={int(r)}, s={int(s)}")
            sat = True
            break

    if not sat:
        print("Not satisfiable")

    # b
    print("b")
    sat = False
    for i in range(16):
        p = (i // 8) % 2 == 1
        q = (i // 4) % 2 == 1
        r = (i // 2) % 2 == 1
        s = i % 2 == 1

        value = ((not p or not q or r)
            and (not p or q or not s)
            and (p or not q or not s)
            and (not p or not r or not s)
            and (p or q or not r)
            and (p or not r or not s))

        if value:
                print(f"Satisfiable for p={int(p)}, q={int(q)}, r={int(r)}, s={int(s)}")
                sat = True
                break

    if not sat:
        print("Not satisfiable")

    # c
    print("c")
    sat = False
    for i in range(16):
        p = (i // 8) % 2 == 1
        q = (i // 4) % 2 == 1
        r = (i // 2) % 2 == 1
        s = i % 2 == 1

        value = ((p or q or r)
          and (p or not q or not s)
          and (q or not r or s)
          and (not p or r or s)
          and (not p or q or not s)
          and (p or not q or not r)
          and (not p or not q or s)
          and (not p or not r or not s))
        
        if value:
                print(f"Satisfiable for p={int(p)}, q={int(q)}, r={int(r)}, s={int(s)}")
                sat = True
                break
    if not sat:
        print("Not satisfiable")

if __name__ == "__main__":
    main()

File: Week_02/PPMH_BT7/BT66/PPMH_BT66.cpp
// a) (p V q V !r) ^ (p V !q V !s) ^ (p V !r V !s) ^ (!p V !q V !s) ^ (p V q V !s)
// b) (!p V !q V r) ^ (!p V q V !s) ^ (p V !q V !s) ^ (!p V !r V !s) ^ (p V q V !s) ^ (q V !r V !s)
// c) (p V q V r) ^ (p V !q V !s) ^ (q V !r V s) ^ (!p V !q V s) ^ (!p V !r V !s)

#include <iostream>
using namespace std;

bool implies(bool a, bool b) {
    return (!a) || b;
}

bool biconditional(bool a, bool b) {
    return a == b;
}

int main() {
    bool p, q, r, s;
    bool value;
    bool sat;

    // a
    cout << "a)" << endl;
    sat = false;
    for(int i = 0; i<16; i++){
        p = (i / 8) % 2;
        q = (i / 4) % 2;
        r = (i / 2) % 2;
        s = i % 2;

        value = ( (p || q || !r)
                && (p || !q || !s)
                && (p || !r || !s)
                && (!p || !q || s)
                && (p || q || !s) );
        if(value){
            cout << "\nSatisfiable for p=" << p << ", q=" << q << ", r=" << r << ", s=" << s;
            sat = true;
            break;
        }
    }
    if(!sat) cout << "\nNot satisfiable";

    // b
    cout << "\n\nb)" << endl;
    sat = false;
    for(int i = 0; i<16; i++){
        p = (i / 8) % 2;
        q = (i / 4) % 2;
        r = (i / 2) % 2;
        s = i % 2;

        value = ( (!p || !q || r)
               && (!p || q || !s)
               && (p || !q || !s)
               && (!p || !r || !s)
               && (p || q || !r)
               && (p || !r || !s) );
        if(value){
            cout << "\nSatisfiable for p=" << p << ", q=" << q << ", r=" << r << ", s=" << s;
            sat = true;
            break;
        }
    }
    if(!sat) cout << "\nNot satisfiable";

    // c
    cout << "\n\nc)" << endl;
    sat = false;
    for(int i = 0; i<16; i++){
        p = (i / 8) % 2;
        q = (i / 4) % 2;
        r = (i / 2) % 2;
        s = i % 2;

        value = ( (p || q || r)
               && (p || !q || !s)
               && (q || !r || s)
               && (!p || r || s)
               && (!p || q || !s)
               && (p || !q || !r)
               && (!p || !q || s)
               && (!p || !r || !s) );

        if(value){
            cout << "\nSatisfiable for p=" << p << ", q=" << q << ", r=" << r << ", s=" << s;
            sat = true;
            break;
        }
    }
    if(!sat) cout << "\nNot satisfiable";
    cout << endl;
    return 0;
}